r/tests/testthat/test-backwards-compatibility.R

```diff
@@ -0,0 +1,159 @@
+  expect_equal(df, example_with_metadata)
+  expect_identical(dim(df), c(1L, 4L))
+
+  expect_equal(
+    attributes(df),
+    list(
+      names = letters[1:4],
+      row.names = 1L,
+      top_level = list(field_one = 12, field_two = "more stuff"),
+      class = c("tbl_df", "tbl", "data.frame"))
+  )
+  expect_equal(attributes(df$a), list(class = "special_string"))
+  expect_null(attributes(df$b))
+  expect_equal(
+    attributes(df$c),
+    list(
+      row.names = 1L,
+      names = c("c1", "c2", "c3"),
+      class = c("tbl_df", "tbl", "data.frame")
+    )
+  )
+  expect_null(attributes(df$d))
+})
+
+test_that("reading a known Parquet file to dataframe with 1.0.1", {
+  skip_if_not_available("snappy")
+  pq_file <- test_path("golden-files/data-arrow_1.0.1.parquet")
+
+  df <- read_parquet(pq_file)
+  # 1.0.1 didn't save top-level metadata, so we need to remove it.
+  example_with_metadata_sans_toplevel <- example_with_metadata
+  attributes(example_with_metadata_sans_toplevel)$top_level <- NULL
+  expect_equal(df, example_with_metadata_sans_toplevel)
+  expect_identical(dim(df), c(1L, 4L))
+
+  expect_equal(
+    attributes(df),
+    list(
+      names = letters[1:4],
+      row.names = 1L,
+      class = c("tbl_df", "tbl", "data.frame"))
+  )
+  expect_equal(attributes(df$a), list(class = "special_string"))
+  expect_null(attributes(df$b))
+  expect_equal(
+    attributes(df$c),
+    list(
+      row.names = 1L,
+      names = c("c1", "c2", "c3"),
+      class = c("tbl_df", "tbl", "data.frame")
+    )
+  )
+  expect_null(attributes(df$d))
+})
+
+for (comp in c("lz4", "uncompressed", "zstd")) {
+  # write_feather(example_with_metadata, test_path("golden-files/data-arrow_2.0.0_lz4.feather"), compression = "lz4")
+  # write_feather(example_with_metadata, test_path("golden-files/data-arrow_2.0.0_uncompressed.feather"), compression = "uncompressed")
+  # write_feather(example_with_metadata, test_path("golden-files/data-arrow_2.0.0_zstd.feather"), compression = "zstd")
+  test_that("reading a known Feather file to dataframe with 2.0.0", {
+    skip_if_not_available(comp)
+    feather_file <- test_path(paste0("golden-files/data-arrow_2.0.0_", comp,".feather"))
+
+    df <- read_feather(feather_file)
+    expect_equal(df, example_with_metadata)
+    expect_identical(dim(df), c(1L, 4L))
+
+    expect_equal(
+      attributes(df),
+      list(
+        names = letters[1:4],
+        row.names = 1L,
+        top_level = list(field_one = 12, field_two = "more stuff"),
+        class = c("tbl_df", "tbl", "data.frame"))
+    )
+    expect_equal(attributes(df$a), list(class = "special_string"))
+    expect_null(attributes(df$b))
+    expect_equal(
+      attributes(df$c),
+      list(
+        row.names = 1L,
+        names = c("c1", "c2", "c3"),
+        class = c("tbl_df", "tbl", "data.frame")
+      )
+    )
+    expect_null(attributes(df$d))
+  })
+
+  test_that("reading a known Feather file to dataframe with 1.0.1", {
```

Comment:
Shouldn't this be failing on master per ARROW-10850?

Reply:
It looks like 1.0.1 works fine, but earlier versions (e.g. 0.17.0) saves files with the wrong case and does fail, but 1.0.1 (and 1.0.0) doesn't seem to have this issue. I need to dig more to inspect the files saved by 1.0.1 / through the source to find if there was a `tolower()` or the like that was introduced after 0.17

Reply:
Indeed, AFAIK, the bug only occurs for files written with 0.17, and when reading with 2.0.0. 
In Arrow 0.17, we had experimental compression support specifically for feather (using some metadata string), while starting from Arrow 1.0, compression is part of the official format spec. So it's only those older pre-1.0.0 files that are the problem. And it only started failing to read them in Arrow 2.0, because we simply changed how the compressions are coded internally from lower case (as it was stored in the feather file) to upper case, in an unrelated commit (https://github.com/apache/arrow/commit/3694794bdfd0677b95b8c95681e392512f1c9237#r43851633), forgetting the compatibility issue with older feather files.

Reply:
Thanks for confirming this. I went searching on Friday and hadn't yet fully reconstructed that time log from git history :)